qiskit/providers/backend_compat.py

```diff
@@ -0,0 +1,315 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=no-value-for-parameter,unused-wildcard-import,wildcard-import
+
+"""Backend abstract interface for providers."""
+
+from typing import List, Iterable, Any, Dict, Optional
+
+from qiskit.providers.backend import BackendV1, BackendV2
+from qiskit.transpiler.target import Target, InstructionProperties
+from qiskit.providers.backend import QubitProperties
+from qiskit.utils.units import apply_prefix
+from qiskit.circuit.library.standard_gates import *
+from qiskit.circuit.parameter import Parameter
+from qiskit.circuit.gate import Gate
+from qiskit.circuit.delay import Delay
+from qiskit.circuit.measure import Measure
+from qiskit.circuit.reset import Reset
+from qiskit.providers.models.backendconfiguration import BackendConfiguration
+from qiskit.providers.models.backendproperties import BackendProperties
+from qiskit.providers.models.pulsedefaults import PulseDefaults
+from qiskit.providers.options import Options
+from qiskit.providers.exceptions import BackendPropertyError
+
+
+def convert_to_target(
+    configuration: BackendConfiguration,
+    properties: BackendProperties = None,
+    defaults: PulseDefaults = None,
+    custom_name_mapping: Optional[Dict[str, "Operation"]] = None,
+) -> Target:
+    """Uses configuration, properties and pulse defaults
+    to construct and return Target class.
+    """
+    # Standard gates library mapping, multicontrolled gates not included since they're
+    # variable width
+    name_mapping = {
```

Comment:
Can we dynamically search the namespace or define this mapping in `qiskit.circuit.library.standard_gates`? My concern is when someone add new gate class, one could forget to add the gate here.

Reply:
Doing it dynamically is tricky for two reasons, first we need the name which is a runtime parameter for most of the gate classes so we don't know what it just from the class. The second reason is we need an initialized gate object for the output of the mapping which is hard to do without inspecting the code to know how many parameters a gate takes.

I'll move it to `qiskit.circuit.library.standard_gates` I do wonder about the import overhead though of iniitalizing a bunch of gate objects in a path that is almost certainly imported. But I guess this new module will get pulled in by default too so there isn't probably any difference (I'll also make it lazy loaded just in case there is an issue).

Reply:
Moved to the standard gates module in: https://github.com/Qiskit/qiskit-terra/pull/8759/commits/f2803ad2bebfac1780ad5b6ee4563cc6a58ee79e